### PR TITLE
Implement stack prescale logic

### PR DESCRIPTION
**Pull request by mikkeloscar**
This is a rough first draft for adding stack prescale logic to the controller. I have tried to pull out the scaling related code into what I call `TrafficReconciler` and implemented a `SimpleTrafficReconciler` which uses the same logic as before and added a `PrescaleTrafficReconciler` which implements the prescale logic.

Suggestions on how to improve this are very welcome. Currently it's passing a lot of information down because a lot of information is needed to decide how to do the traffic switching.

## TODO

* [x] Add a lot more tests
* [x] Test for real and see if behavior is as expected.
* [x] Document the prescaling logic.

### CODE REVIEW

File: controller/prescale_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	prescaleAnnotationKey = "stacksetstacks.zalando.org/prescale-replicas"
+)
+
+type PrescaleTrafficReconciler struct {
+	base TrafficReconciler
```
Comment (linki) [inline controller/prescale_reconciler.go]: Seems like `base` isn't used after all.

**Comment by linki** [comment]
The more code the less comments..
Looking forward to seeing this in action.
lgtm 👍

**Comment by mikkeloscar** [comment]
Added tests, fixed bugs we found during manual testing and added docs.

PTAL!

**Comment by linki** [comment]
:+1:

**Comment by mikkeloscar** [comment]
:+1:

**Pull request merged**